libbeat/publisher/queue/proxy/queue_test.go

```diff
@@ -0,0 +1,220 @@
+	}
+
+	assert.Equal(t, 0, listener.ackedCount, "No batches have been acked yet")
+	for _, batch := range batches {
+		batch.Done()
+	}
+	assert.NoError(t, listener.waitForTotalACKs(PRODUCER_COUNT, time.Second))
+}
+
+func TestOutOfOrderACK(t *testing.T) {
+	logger := logp.NewLogger("proxy-queue-tests")
+
+	// Create a proxy queue where each batch is at most 2 events
+	testQueue := NewQueue(logger, Settings{BatchSize: 2})
+	defer testQueue.Close()
+
+	listener := newTestACKListener()
+	producer := testQueue.Producer(queue.ProducerConfig{
+		ACK: listener.ACK,
+	})
+
+	const BATCH_COUNT = 10
+	batches := []queue.Batch{}
+	for i := 0; i < BATCH_COUNT; i++ {
+		// Publish two events
+		_, success := producer.Publish(0)
+		assert.True(t, success, "Publish should succeed")
+		_, success = producer.Publish(0)
+		assert.True(t, success, "Publish should succeed")
+
+		// Consume a batch, which should contain the events we just published
+		batch, err := testQueue.Get(0)
+		assert.NoError(t, err)
+		batch.FreeEntries()
+		assert.Equal(t, 2, batch.Count())
+
+		batches = append(batches, batch)
+	}
+
+	// Acknowledge all except the first batch
+	for _, batch := range batches[1:] {
+		batch.Done()
+	}
+	// Make sure that no ACKs come in even if we wait a bit
+	err := listener.waitForTotalACKs(1, 50*time.Millisecond)
+	assert.Error(t, err, "No ACK callbacks should have been called yet")
+
+	// ACKing the first batch should unblock all the rest
+	batches[0].Done()
+	assert.NoError(t, listener.waitForTotalACKs(BATCH_COUNT*2, time.Second))
+}
+
+func newTestACKListener() *testACKListener {
+	return &testACKListener{
+		updateChan: make(chan struct{}, 1),
+	}
+}
+
+// ACK should be provided to the queue producer. It can be safely called from
+// multiple goroutines.
+func (l *testACKListener) ACK(count int) {
+	l.Lock()
+	l.ackedCount += count
+	if l.waiting {
+		// If waitFortotalACKs is waiting on something, wake it up so it can retry.
+		l.waiting = false
+		l.updateChan <- struct{}{}
+	}
+	l.Unlock()
+}
+
+// flush should be called on timeout, to clear updateChan if needed.
+func (l *testACKListener) flush() {
+	l.Lock()
+	select {
+	case <-l.updateChan:
+	default:
+	}
+	l.waiting = false
+	l.Unlock()
+}
+
+// waitForTotalACKs waits until the specified number of total ACKs have been
+// received, or the timeout interval is exceeded. It should only be called
+// from a single goroutine at once.
+func (l *testACKListener) waitForTotalACKs(targetCount int, timeout time.Duration) error {
+	timeoutChan := time.After(timeout)
+	for {
+		l.Lock()
+		if l.ackedCount >= targetCount {
+			l.Unlock()
+			return nil
+		}
+		// Not enough ACKs have been sent yet, so we have to wait.
+		l.waiting = true
+		l.Unlock()
+		select {
+		case <-l.updateChan:
+			// New ACKs came in, retry
+			continue
+		case <-timeoutChan:
+			l.flush()
+			return fmt.Errorf("timed out waiting for acknowledgments: have %d, wanted %d", l.ackedCount, targetCount)
+		}
+	}
+}
```

Comment:
Can you add a few more tests around shutdown, so if queue is shutting down we don't publish etc.

Reply:
I will think about what tests might be appropriate, but:
- there's on some level a fundamental nondeterminism: it's always possible for the shutdown signal to come in simultaneously with a publish request, and what breaks the tie isn't which was sent first but which the queue loop _receives_ first (which we can't tell from outside). We can be certain that no publish requests will go through _after_ `Close` returns (or... at least now we can, since I just added the mistakenly-omitted `Wait` call now in response to this comment 😅) but we can't have any guarantees based on the close channel alone, or while `Close` is still in progress. But, I can add tests to make sure that write-after-close fails. However, sadly:
- `queue.Close` is never called during Beats shutdown, so it's nice to know that it would work correctly if we used it, but as of right now none of the queues are ever shut down properly.

Reply:
A test that shows write after close fails would satisfy me.  I think that goes a long way in exercising that code path.